Add unit tests, e2e test to parser service

TueeNguyen opened the pull request:
<!--
Thanks for sending a pull request!
- if this is your first time, please ensure you have read through our contributor guide: https://github.com/Seneca-CDOT/telescope/blob/master/docs/CONTRIBUTING.md
-->

## Issue This PR Addresses
Part of #2111 
<!--
1. Automatically close the issue when this PR is merged
    USAGE: Fixes #<issue number>
2. If your PR addresses an issue but does not close it
    USAGE: #<issue number> <reason>(i.e. This issue was worked on by @user and myself and his PR should close the issue.)
-->

## Type of Change

<!-- bug fix, feature, documentation, UI, etc. -->

- [ ] **Bugfix**: Change which fixes an issue
- [ ] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
- Most tests are copied from `test/`, I fixed the some imports and added a few cases 
- e2e test is completed

<!-- Please add a detailed description of what this PR does and why it is needed -->

## Steps to test the PR
* Remove redis cache `rm -rf redis-data` or else it'll take very long and the e2e test will timeout
* pnpm install
* pnpm services:start elasticsearch redis traefik posts
* pnpm jest:e2e src/api/parser

**HOWEVER**
If you
```
pnpm services:start elasticsearch posts redis traefik
pnpm jest:e2e src/api/parser
```
It'll error out because the change in #3248 make all e2e tests to wait for Postgre. So you should `pnpm services:start` and run the test. 

This could be annoying when we have more e2e tests and want to run only a specific e2e test in development mode
<!-- Please add steps to build the changes in your PR locally so that peers can test your PR
    Example:
    - `npm install`
    - Go to '...'
    - Click on '....'
    - Scroll down to '....'
    - See error

    Please remember, the more clear you are, the faster your PR will be reviewed and merged.
 -->

## Checklist

<!-- Before submitting a PR, address each item -->

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

JerryHue commented on src/api/parser/test/feed.test.js:
```diff
@@ -0,0 +1,280 @@
+      expect(feed.user).toEqual(data.user);
+    });
+
+    test('Feed.byId() for an invalid url should return null', async () => {
```
```suggestion
    test('Feed.byUrl() for an invalid url should return null', async () => {
```

TueeNguyen commented on src/api/parser/test/url-parser.test.js:
```diff
@@ -0,0 +1,48 @@
+  const expectedReturn2 = 'http://127.0.0.1/';
+
+  test('urlParser with double ports', () => {
+    const result = urlParser(testUrlWithPort, testport);
+    expect(result).toBe(expectedReturn1);
+  });
+
+  test('urlParser with 1 string port', () => {
+    const result = urlParser(testUrlWithoutPort, testport);
+    expect(result).toBe(expectedReturn1);
+  });
+
+  test('urlParser with invalid url', () => {
+    const result = urlParser('', testport);
+    expect(result).toBe(null);
+  });
+
+  test('urlParser with url with no port and no port', () => {
+    const result = urlParser(testUrlWithoutPort);
+    expect(result).toBe(expectedReturn2);
+  });
+
+  test('urlPaser with port being a null value', () => {
+    const result = urlParser(testUrlWithoutPort, null);
+    expect(result).toBe(expectedReturn2);
+  });
+
+  test('urlPaser with port being an integer', () => {
+    const result = urlParser(testUrlWithoutPort, Number(testport));
+    expect(result).toBe(expectedReturn1);
+  });
+
+  test('urlPaser with port being an invalid string', () => {
+    const result = urlParser(testUrlWithoutPort, 'invalid');
+    expect(result).toBe(expectedReturn2);
+  });
+  test('urlPaser with an out of range port', () => {
+    const result = urlParser(testUrlWithoutPort, 999999);
+    expect(result).toBe(expectedReturn2);
+  });
```
Interesting, will look into it

TueeNguyen replied:
Thanks for this tip, I find it useful or repeating tests

TueeNguyen commented on src/api/parser/jest.config.js:
```diff
@@ -3,6 +3,7 @@ const baseConfig = require('../../../jest.config.base');
 module.exports = {
   ...baseConfig,
   rootDir: '../../..',
+  setupFiles: ['<rootDir>/src/api/parser/jest.setup.js', '<rootDir>/test/jest.setup.js'],
```
@rclee91 what do you think?

rclee91 replied:
There is a `node-fetch` mock In the [/test/jest.setup.js](https://github.com/Seneca-CDOT/telescope/blob/master/test/jest.setup.js) plus some other configs. 
Parser currently uses `node-fetch` from Satellite, but it's also just the regular `node-fetch`
But true, if the backend will end up getting ripped one day, we should also move the setup stuff from `/test` to the the setup files in Parser tests.

TueeNguyen commented on src/api/parser/test/feed-queue.test.js:
```diff
@@ -0,0 +1,25 @@
+ */
+describe('Testing feedQueue', () => {
+  test('able to add a job to the feed queue', () =>
+    feedQueue.add('http://ajhooper.blogspot.com/feeds/posts/default'));
```
Fixed!

TueeNguyen commented on package.json:
```diff
@@ -18,7 +18,7 @@
     "test": "pnpm jest",
     "jest": "jest -c jest.config.js --",
-    "jest:e2e": "jest -c jest.config.e2e.js --",
+    "jest:e2e": "jest -c jest.config.e2e.js --forceExit",
```
I added --forceExit because after e2e tests are finished, there are a few open Redis connections and open fetch or logger( this I'm not sure). Jest will be hung and not exit.
https://github.com/Seneca-CDOT/telescope/blob/95dde6cb4d852d9967dc312980a3cb1f1305e334/src/api/parser/src/lib/queue.js#L9-L10

TueeNguyen commented:
My main concern now is with `logger` and `setInterval` hanging Jest. In terms of Redis, I've just gotta close the connections

JerryHue commented:
Would it be possible to turn off logging during testing? I am not sure if Satellite has that option, but if we can add it, we wouldn't have to worry about managing anything related to that, right?

TueeNguyen commented:
@JerryHue Nice idea! There's still the `setInterval()`
```
●  TCPWRAP

      147 |
      148 |   const connectivity = new Promise((resolve) => {
    > 149 |     intervalId = setInterval(async () => {
          |                  ^
      151 |         await checkConnection();
      152 |         resolve();

      at src/api/parser/src/utils/indexer.js:149:18
      at waitOnReady (src/api/parser/src/utils/indexer.js:148:24)
      at Object.start (src/api/parser/src/feed/worker.js:30:11)
      at processFeeds (src/api/parser/test/e2e/parser-flow.test.js:57:14)
      at src/api/parser/test/e2e/parser-flow.test.js:61:3
      at TestScheduler.scheduleTests (node_modules/.pnpm/@jest+core@27.5.1/node_modules/@jest/core/build/TestScheduler.js:333:13)
      at runCLI (node_modules/.pnpm/@jest+core@27.5.1/node_modules/@jest/core/build/cli/index.js:173:3)
```

humphd commented:
> Would it be possible to turn off logging during testing? I am not sure if Satellite has that option, but if we can add it, we wouldn't have to worry about managing anything related to that, right?

One thing that we could try is altering how Satellite works in testing/CI.  An easy way would be to add some `if` checks for `process.env.NODE_ENV === 'test'` or something.  In "test" mode vs. "production" or "development" we can alter how we setup various things.

Since we own Satellite, we can bend it into any shape we want to support our use case.

humphd commented:
Also, with timer's you can `unref()` them so they don't block the process, see https://httptoolkit.tech/blog/unblocking-node-with-unref/#timeoutunref

TueeNguyen commented:
I used `unref()` but it still hangs.

TueeNguyen commented:
Using new Satellite's `stop()` function still hangs Jest, it seems like the Promises returned from `stop()` don't resolve or reject.

humphd commented:
I pulled your branch and ran it locally.  I had to do two things:

1. rebase onto `master` to get the `src/satellite` files to exist (mine only had `src/satellite/node_modules`).
2. add the parser tests to the root Jest config:

```diff
diff --git a/jest.config.js b/jest.config.js
index 94a2bb7b..aff86c64 100644
--- a/jest.config.js
+++ b/jest.config.js
@@ -16,5 +16,6 @@ module.exports = {
     '<rootDir>/src/api/posts/jest.config.js',
     '<rootDir>/src/api/search/jest.config.js',
     '<rootDir>/src/api/sso/jest.config.js',
+    '<rootDir>/src/api/parser/jest.config.js',
   ],
 };
```

Once I did this, I ran the parser tests with:

```sh
$ pnpm test parser
```

This ran all kinds of things and hung.  Looking at the tests, I noticed there were tests running in `tests/parser` (i.e., the old parser tests):

```
 PASS  test/wiki-feed-parser.test.js
  ✓ Testing wiki-feed-parser.parseData (243 ms)
...
 PASS  test/url-parser.test.js
  url-parser tests
    ✓ urlParser with double ports (1 ms)
    ✓ urlParser with 1 string port
    ✓ urlParser with invalid url (1 ms)
    ✓ urlParser with url with no port and no port
    ✓ urlPaser with port being a null value
    ✓ urlPaser with port being an integer
    ✓ urlPaser with port being an invalid string (3 ms)
    ✓ urlPaser with an out of range port (1 ms)
...
 PASS  test/text-parser.test.js
  text-parser tests
    ✓ textParser() with doctype (37 ms)
    ✓ textParser() with HTML body (2 ms)
    ✓ textParser() with minimal document fragment (1 ms)
    ✓ textParser() with real blog post (8 ms)
...
A worker process has failed to exit gracefully and has been force exited. This is likely caused
by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks.
Active timers can also cause this, ensure that .unref() was called on them.
```

If I remove these tests from the list of running tests, I get no leaked handles:

```
$ pnpm test api/parser
...

Test Suites: 17 passed, 17 total
Tests:       145 passed, 145 total
Snapshots:   0 total
Time:        1.957 s, estimated 2 s
Ran all test suites matching /api\/parser/i.
```

So I'd say your code is working, and you need to remove the old broken tests :)

rclee91 commented:
I know the unit tests were working fine. I think the `stop()` was mostly for e2e tests. Though, maybe @TueeNguyen can elaborate more on that

humphd commented:
e2e tests also pass for me:

```sh
$ pnpm jest:e2e api/parser
 PASS  src/api/parser/test/e2e/parser-flow.test.js
  Testing parser service's flow
    ✓ Parser should process 2 valid feeds (2464 ms)
    ✓ Parser should process 2 invalid feeds (480 ms)
    ✓ Parser should process 1 Youtube Feed

Test Suites: 1 passed, 1 total
Tests:       3 passed, 3 total
Snapshots:   0 total
Time:        4.549 s
```

humphd commented:
However, after the tests end, I do see:

```
[ioredis] Unhandled error event: Error: connect ECONNREFUSED 127.0.0.1:6379
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1157:16)
...
```

This turns out to be:

```
ps aux | grep node | grep telescope
humphd           85835   0.0  0.3 410248688 101920 s001  S     9:35pm   0:02.15 /Users/humphd/.nvm/versions/node/v16.14.0/bin/node /Users/humphd/repos/telescope/node_modules/.pnpm/bull@3.29.3/node_modules/bull/lib/process/master.js
```

So we aren't properly shutting down the Bull queue in the parser (outside of Satellite).

TueeNguyen commented:
If you pulled the changes from this PR, I have add `--forceExit` as a temporary solution to Jest hanging. Did you run it without `--forceExit`. 
I'm going back out that change so testing this will be more accurate

humphd commented:
For shutting down the feed queue worker, see https://docs.bullmq.io/guide/workers/graceful-shutdown.

TueeNguyen commented:
I've tried it before, doing `feedQueue.close()` doesn't seem to work for me. I've added `feedQueue.client.quit()` which seems to close Redis connection.

I've added this in my latest commit
```js
afterAll(() => Promise.all[(feedQueue.close(), feedQueue.client.quit())]);
```

humphd commented on src/api/parser/test/e2e/parser-flow.test.js:
```diff
@@ -0,0 +1,105 @@
+      logger.error({ error }, 'Unable to invalidate feed')
+    )
+  );
+  feedWorker.start();
```
This needs to get shutdown too in your `afterAll`.

TueeNguyen commented:
I've made parser test to [`forceExit`](https://jestjs.io/docs/cli#--forceexit) by 
* Adding the script `jest:e2e-all` that runs other e2e tests normally (we only have sso and parser for now) and parser's e2e test separately (forceExit only this)
* Change our `e2e-tests-ci.yml` to run `pnpm jest:e2e-all` instead of `pnpn jest:e2e`

humphd commented on package.json:
```diff
@@ -22,6 +22,7 @@
     "test": "pnpm jest",
     "jest": "jest -c jest.config.js --",
     "jest:e2e": "jest -c jest.config.e2e.js --",
+    "jest:e2e-all": "cross-env e2e_all=1 jest -c jest.config.e2e.js -- && cross-env parser_e2e=1 jest -c jest.config.e2e.js --forceExit -- ",
```
I don't understand this change.  First, can we move the environment variables into the Jest setup vs. here?  Also, I think it's fine to add `--forceExit` to the regular `jest:e2e` call.

TueeNguyen replied:
ahh that's even better then

humphd commented on package.json:
```diff
@@ -99,6 +100,7 @@
     "@types/jest": "27.4.1",
     "@vercel/node": "1.13.0",
     "babel-preset-next": "1.4.0",
+    "cross-env": "7.0.3",
```
We just got rid of this, can we avoid adding it back?

humphd commented on pnpm-lock.yaml:
```diff
@@ -123,10 +124,11 @@ importers:
       '@babel/preset-react': 7.16.7_@babel+core@7.17.8
       '@babel/preset-typescript': 7.16.7_@babel+core@7.17.8
-      '@senecacdot/eslint-config-telescope': 1.1.0_eslint@7.32.0
+      '@senecacdot/eslint-config-telescope': link:tools/eslint
```
We want to avoid this.  Rebasing should fix?

TueeNguyen replied:
I just rebased, I probably missed this

TueeNguyen commented:
* Should I install the 2 dependencies `jest-fetch-mock` and `nock`?
* I think it's ok for expect to be inside of a test block, I can open a PR to patch this quickly
```
@senecacdot/parser-service:lint: /home/runner/work/telescope/telescope/src/api/parser/jest.setup.js
@senecacdot/parser-service:lint:   1:15  error  'jest-fetch-mock' should be listed in the project's dependencies. Run 'npm i -S jest-fetch-mock' to add it  import/no-extraneous-dependencies
@senecacdot/parser-service:lint: 
@senecacdot/parser-service:lint: /home/runner/work/telescope/telescope/src/api/parser/test/fixtures.js
@senecacdot/parser-service:lint:   1:14  error  'nock' should be listed in the project's dependencies. Run 'npm i -S nock' to add it  import/no-extraneous-dependencies
@senecacdot/parser-service:lint: 
@senecacdot/parser-service:lint: /home/runner/work/telescope/telescope/src/api/parser/test/post.test.js
@senecacdot/parser-service:lint:   190:7  error  Expect must be inside of a test block  jest/no-standalone-expect
@senecacdot/parser-service:lint:   191:7  error  Expect must be inside of a test block  jest/no-standalone-expect
@senecacdot/parser-service:lint:   [298](https://github.com/Seneca-CDOT/telescope/runs/5646948164?check_suite_focus=true#step:5:298):7  error  Expect must be inside of a test block  jest/no-standalone-expect
@senecacdot/parser-service:lint:   [299](https://github.com/Seneca-CDOT/telescope/runs/5646948164?check_suite_focus=true#step:5:299):7  error  Expect must be inside of a test block  jest/no-standalone-expect
```

humphd reviewed (changes requested):
```
@senecacdot/parser-service:lint: /home/runner/work/telescope/telescope/src/api/parser/jest.setup.js
@senecacdot/parser-service:lint:   1:15  error  'jest-fetch-mock' should be listed in the project's dependencies. Run 'npm i -S jest-fetch-mock' to add it  import/no-extraneous-dependencies
@senecacdot/parser-service:lint: 
@senecacdot/parser-service:lint: /home/runner/work/telescope/telescope/src/api/parser/test/fixtures.js
@senecacdot/parser-service:lint:   1:14  error  'nock' should be listed in the project's dependencies. Run 'npm i -S nock' to add it  import/no-extraneous-dependencies
@senecacdot/parser-service:lint: 
@senecacdot/parser-service:lint: /home/runner/work/telescope/telescope/src/api/parser/test/post.test.js
@senecacdot/parser-service:lint:   190:7  error  Expect must be inside of a test block  jest/no-standalone-expect
@senecacdot/parser-service:lint:   191:7  error  Expect must be inside of a test block  jest/no-standalone-expect
@senecacdot/parser-service:lint:   [298](https://github.com/Seneca-CDOT/telescope/runs/5648861362?check_suite_focus=true#step:5:298):7  error  Expect must be inside of a test block  jest/no-standalone-expect
@senecacdot/parser-service:lint:   [299](https://github.com/Seneca-CDOT/telescope/runs/5648861362?check_suite_focus=true#step:5:299):7  error  Expect must be inside of a test block  jest/no-standalone-expect
@senecacdot/parser-service:lint: 
@senecacdot/parser-service:lint: ✖ 6 problems (6 errors, 0 warnings)
@senecacdot/parser-service:lint: 
@senecacdot/parser-service:lint: Rule                                | Time (ms) | Relative
@senecacdot/parser-service:lint: :-----------------------------------|----------:|--------:
@senecacdot/parser-service:lint: prettier/prettier                   |  6266.722 |    64.9%
@senecacdot/parser-service:lint: react/no-did-update-set-state       |   272.471 |     2.8%
@senecacdot/parser-service:lint: import/no-unresolved                |   203.901 |     2.1%
@senecacdot/parser-service:lint: react/destructuring-assignment      |   184.235 |     1.9%
@senecacdot/parser-service:lint: react/no-deprecated                 |   136.604 |     1.4%
@senecacdot/parser-service:lint: react/void-dom-elements-no-children |   126.635 |     1.3%
@senecacdot/parser-service:lint: react/jsx-no-bind                   |   117.262 |     1.2%
@senecacdot/parser-service:lint: import/extensions                   |   105.112 |     1.1%
@senecacdot/parser-service:lint: jest/no-standalone-expect           |   102.808 |     1.1%
@senecacdot/parser-service:lint: react/no-this-in-sfc                |   102.142 |     1.1%
@senecacdot/parser-service:lint:  ELIFECYCLE  Command failed with exit code 1.
@senecacdot/parser-service:lint:  ELIFECYCLE  Command failed with exit code 1.
@senecacdot/parser-service:lint: Error: command finished with error: command (src/api/parser) pnpm run lint exited (1)
command (src/api/parser) pnpm run lint exited (1)
```

humphd reviewed (commented):
...And just as I click "submit review" I remember that our CI workflows won't run what's submitted in PRs, only what we do on master.

Can you break the change to `.github/workflows/e2e-tests-ci.yml` out into a separate PR we can land right away, and then you can rebase on it to run these with ES.

TueeNguyen commented:
@humphd, I reverted the CI change.

humphd commented:
Please do a PR for the CI change on its own, or these will never pass.

TueeNguyen commented:
@humphd, it passed all checks 🎆

humphd reviewed (approved):
We should look into why Playwright takes so long to setup in CI.  They have a Docker image we can use, which might be faster.

This is good to go I think!

TueeNguyen commented:
For anyone who's testing this out. If you
```
pnpm services:start elasticsearch posts redis traefik
pnpm jest:e2e src/api/parser
```
It'll error out because the change in #3248 make all e2e tests to wait for Postgre. So you should start all services and run the test. 

This could be annoying when we have more e2e tests and want to run only a specific e2e test in development mode

HyperTHD commented on src/api/parser/test/sanitize-html.test.js:
```diff
@@ -0,0 +1,151 @@
+  });
+
+  test('twitch.tv embedded content should not be removed', () => {
+    const data = sanitizeHTML('<iframe src="https://www.twitch.tv/0pensrc"></iframe>');
```
Embedded twitch content doesn't come from here, should be from player.twitch.tv
Thought I added the test for this here though, strange it's not here

TueeNguyen replied:
Thanks for pointing it out, maybe it got messed up when rebasing

HyperTHD reviewed (changes requested):
Just one last thing and this is good

Kevan-Y reviewed (approved):
Works tested on local.

TueeNguyen commented:
@HyperTHD, I think it's good to go but I probably need your last review on the sanitize file

HyperTHD reviewed (approved):
LGTM @TueeNguyen 
Just make sure you rebase prior to merging
